src/newgrp.c

```diff
@@ -821,12 +826,20 @@ int main (int argc, char **argv)
 	closelog ();
 #ifdef WITH_AUDIT
 	if (NULL != group) {
-		SNPRINTF(audit_buf, "changing new-group=%s", group);
+		char enc_group[(GROUP_NAME_MAX_LENGTH*2)+1];
+		int len = strnlen(group, sizeof(enc_group)/2);
+		if (audit_value_needs_encoding(group, len)) {
+			SNPRINTF(audit_buf,
+					 "changing new_group=%s",
+					 audit_encode_value(enc_group, group, len));
+		} else {
+			SNPRINTF(audit_buf, "changing new_group=%s", group);
+		}
 		audit_logger (AUDIT_CHGRP_ID, Prog,
-		              audit_buf, NULL, getuid (), 0);
+		              audit_buf, NULL, getuid (), SHADOW_AUDIT_FAILURE);
 	} else {
 		audit_logger (AUDIT_CHGRP_ID, Prog,
-		              "changing", NULL, getuid (), 0);
+		              "changing", NULL, getuid (), SHADOW_AUDIT_FAILURE);
 	}
```

Comment:
This looks like open-coding the function you're adding in the second commit.  Can we encapsulate it in a function?  Why is this open-coded here?

Reply:
The original patch predates my introduction to the professional software world. I tried updating the code to the upstream standards, but I was unable to do it completely.

I have updated the three instances where this type of code was present to use `audit_logger_with_group()` instead. I had to add an additional argument to the function so that everything matched correctly. Take a look at the new code.

Reply:
Makes sense.  Thanks!